### ISSUE TITLE
raft engine seems dead lock

### ISSUE THREAD

**Issue by BusyJay** [tikv/raft-engine#71]
I'm loading data into TiKV, and one of the node somehow is stuck. The stacktrace looks like following.

```
  70   Thread 0x7f3faa67e700 (LWP 22442) "background-1" 0x00007f3fb3ac154d in __lll_lock_wait () from /lib64/libpthread.so.0
  32   Thread 0x7f3f837df700 (LWP 22496) "raftstore-4-0" 0x00007f3fb3ac154d in __lll_lock_wait () from /lib64/libpthread.so.0
  31   Thread 0x7f3f835de700 (LWP 22497) "raftstore-4-1" 0x00007f3fb3ac154d in __lll_lock_wait () from /lib64/libpthread.so.0

Thread 70 (Thread 0x7f3faa67e700 (LWP 22442)):
#0  0x00007f3fb3ac154d in __lll_lock_wait () from /lib64/libpthread.so.0
#1  0x00007f3fb3abce9b in _L_lock_883 () from /lib64/libpthread.so.0
#2  0x00007f3fb3abcd68 in pthread_mutex_lock () from /lib64/libpthread.so.0
#3  0x000055dd465fc5fa in raft_engine::purge::PurgeManager$LT$E$C$W$C$P$GT$::purge_expired_files::ha4e7409244ce5c8d () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#4  0x000055dd46619659 in _$LT$raft_log_engine..engine..RaftLogEngine$u20$as$u20$engine_traits..raft_engine..RaftEngine$GT$::purge_expired_files::h6267e943f2a6fbfa () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#5  0x000055dd46dfe7ba in raftstore::store::worker::raftlog_gc::Runner$LT$EK$C$ER$C$R$GT$::flush::h2c1d38c097dfcf3d () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#6  0x000055dd46b7367c in _$LT$raftstore..store..worker..raftlog_gc..Runner$LT$EK$C$ER$C$R$GT$$u20$as$u20$tikv_util..worker..pool..Runnable$GT$::run::h64bbe6d2c6cb9fa9 () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#7  0x000055dd46d5e7a3 in _$LT$core..future..from_generator..GenFuture$LT$T$GT$$u20$as$u20$core..future..future..Future$GT$::poll::h3b0d350c770ce605 () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#8  0x000055dd47b2b881 in _$LT$yatp..task..future..Runner$u20$as$u20$yatp..pool..runner..Runner$GT$::handle::hc59a7a64f171a303 ()
#9  0x000055dd46ca7cef in yatp::pool::worker::WorkerThread$LT$T$C$R$GT$::run::h33ca1869d4641979 () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#10 0x000055dd46be7c65 in std::sys_common::backtrace::__rust_begin_short_backtrace::h40fdd7e5cdbd32b8 () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#11 0x000055dd470ef5d5 in std::panicking::try::do_call::h19dc3bce6576f8aa () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#12 0x000055dd46bfbe4d in core::ops::function::FnOnce::call_once$u7b$$u7b$vtable.shim$u7d$$u7d$::hc737d6d377959610 () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#13 0x000055dd473389d7 in call_once<(),FnOnce<()>,alloc::alloc::Global> () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/alloc/src/boxed.rs:1546
#14 call_once<(),alloc::boxed::Box<FnOnce<()>, alloc::alloc::Global>,alloc::alloc::Global> () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/alloc/src/boxed.rs:1546
#15 std::sys::unix::thread::Thread::new::thread_start::hfbe13ead469fd0bc () at library/std/src/sys/unix/thread.rs:71
#16 0x00007f3fb3abaea5 in start_thread () from /lib64/libpthread.so.0
#17 0x00007f3fb30c38dd in clone () from /lib64/libc.so.6

Thread 32 (Thread 0x7f3f837df700 (LWP 22496)):
#0  0x00007f3fb3ac154d in __lll_lock_wait () from /lib64/libpthread.so.0
#1  0x00007f3fb3abce9b in _L_lock_883 () from /lib64/libpthread.so.0
#2  0x00007f3fb3abcd68 in pthread_mutex_lock () from /lib64/libpthread.so.0
#3  0x000055dd46600e33 in raft_engine::engine::Engine$LT$E$C$W$C$P$GT$::apply_to_memtable::haf3ea2cc1eee62a5 () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#4  0x000055dd46601c29 in raft_engine::engine::Engine$LT$E$C$W$C$P$GT$::write::h775bbd18d2b90939 () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#5  0x000055dd466190e5 in _$LT$raft_log_engine..engine..RaftLogEngine$u20$as$u20$engine_traits..raft_engine..RaftEngine$GT$::consume_and_shrink::h1fa296df8acad832 () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#6  0x000055dd46afc60d in raftstore::store::fsm::store::RaftPoller$LT$EK$C$ER$C$T$GT$::handle_raft_ready::h1ed29fbb2ac0ec36 () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#7  0x000055dd468ee4f0 in _$LT$raftstore..store..fsm..store..RaftPoller$LT$EK$C$ER$C$T$GT$$u20$as$u20$batch_system..batch..PollHandler$LT$raftstore..store..fsm..peer..PeerFsm$LT$EK$C$ER$GT$$C$raftstore..store..fsm..store..StoreFsm$LT$EK$GT$$GT$$GT$::end::h72af608d6cce3a73 () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#8  0x000055dd46b86e6d in batch_system::batch::Poller$LT$N$C$C$C$Handler$GT$::poll::h10865c3245dddabb () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#9  0x000055dd46be9467 in std::sys_common::backtrace::__rust_begin_short_backtrace::h6f5de66a8bb8712e () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#10 0x000055dd470f3c04 in std::panicking::try::do_call::hade66f03795540dd () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#11 0x000055dd46bf91ac in core::ops::function::FnOnce::call_once$u7b$$u7b$vtable.shim$u7d$$u7d$::h4f7954cf4e4bd9b9 () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#12 0x000055dd473389d7 in call_once<(),FnOnce<()>,alloc::alloc::Global> () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/alloc/src/boxed.rs:1546
#13 call_once<(),alloc::boxed::Box<FnOnce<()>, alloc::alloc::Global>,alloc::alloc::Global> () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/alloc/src/boxed.rs:1546
#14 std::sys::unix::thread::Thread::new::thread_start::hfbe13ead469fd0bc () at library/std/src/sys/unix/thread.rs:71
#15 0x00007f3fb3abaea5 in start_thread () from /lib64/libpthread.so.0
#16 0x00007f3fb30c38dd in clone () from /lib64/libc.so.6

Thread 31 (Thread 0x7f3f835de700 (LWP 22497)):
#0  0x00007f3fb3ac154d in __lll_lock_wait () from /lib64/libpthread.so.0
#1  0x00007f3fb3abce9b in _L_lock_883 () from /lib64/libpthread.so.0
#2  0x00007f3fb3abcd68 in pthread_mutex_lock () from /lib64/libpthread.so.0
#3  0x000055dd46600e33 in raft_engine::engine::Engine$LT$E$C$W$C$P$GT$::apply_to_memtable::haf3ea2cc1eee62a5 () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#4  0x000055dd46601c29 in raft_engine::engine::Engine$LT$E$C$W$C$P$GT$::write::h775bbd18d2b90939 () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#5  0x000055dd466190e5 in _$LT$raft_log_engine..engine..RaftLogEngine$u20$as$u20$engine_traits..raft_engine..RaftEngine$GT$::consume_and_shrink::h1fa296df8acad832 () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#6  0x000055dd46afc60d in raftstore::store::fsm::store::RaftPoller$LT$EK$C$ER$C$T$GT$::handle_raft_ready::h1ed29fbb2ac0ec36 () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#7  0x000055dd468ee4f0 in _$LT$raftstore..store..fsm..store..RaftPoller$LT$EK$C$ER$C$T$GT$$u20$as$u20$batch_system..batch..PollHandler$LT$raftstore..store..fsm..peer..PeerFsm$LT$EK$C$ER$GT$$C$raftstore..store..fsm..store..StoreFsm$LT$EK$GT$$GT$$GT$::end::h72af608d6cce3a73 () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#8  0x000055dd46b86e6d in batch_system::batch::Poller$LT$N$C$C$C$Handler$GT$::poll::h10865c3245dddabb () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#9  0x000055dd46be9467 in std::sys_common::backtrace::__rust_begin_short_backtrace::h6f5de66a8bb8712e () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#10 0x000055dd470f3c04 in std::panicking::try::do_call::hade66f03795540dd () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#11 0x000055dd46bf91ac in core::ops::function::FnOnce::call_once$u7b$$u7b$vtable.shim$u7d$$u7d$::h4f7954cf4e4bd9b9 () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/slice/mod.rs:568
#12 0x000055dd473389d7 in call_once<(),FnOnce<()>,alloc::alloc::Global> () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/alloc/src/boxed.rs:1546
#13 call_once<(),alloc::boxed::Box<FnOnce<()>, alloc::alloc::Global>,alloc::alloc::Global> () at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/alloc/src/boxed.rs:1546
#14 std::sys::unix::thread::Thread::new::thread_start::hfbe13ead469fd0bc () at library/std/src/sys/unix/thread.rs:71
#15 0x00007f3fb3abaea5 in start_thread () from /lib64/libpthread.so.0
#16 0x00007f3fb30c38dd in clone () from /lib64/libc.so.6
```

The only configuration change is:
```
[raft-engine]
enable = true
```

**Comment by gotoxu** [comment]
We also met this problem. 

When the raft-engine option is turned on, TiKV will freeze under high writing pressure.

The Recv-Q queue starts to see a backlog.

TiKV no longer outputs any log information

**Comment by tabokie** [comment]
@gotoxu Which TiKV version are you using at that time?
I believe this bug is fixed by #63, and this fix is brought to TiKV in tikv/tikv#10061.

**Comment by gotoxu** [comment]
@tabokie we use 5.0.1 at that time. 

It looks like this bug was fixed in 5.0.2

**Comment by tabokie** [comment]
> @tabokie we use 5.0.1 at that time.
> 
> It looks like this bug was fixed in 5.0.2

@gotoxu I just noticed we haven't bumped the raft-engine version in previous releases (see [Cargo.lock](https://github.com/tikv/tikv/blob/v5.0.2/Cargo.lock) in v5.0.2), you might need to use nightly version to include the fix.
